list.c

```diff
@@ -0,0 +1,66 @@
+#include <stdlib.h>
+#include <stdint.h>
+#include "list.h"
+
+// simple linked list implementation
+
+struct list *list_new(void)
+{
+    return calloc(1, sizeof(struct list));
+}
+
+// removed linked list nodes but does not free their content
+void list_clear(struct list *list)
+{
+    struct list_item *next, *cur = list->head;
+    while (cur != NULL) {
+        next = cur->next;
+        free(cur);
+        cur = next;
+    }
+}
+
+// returns generated linked list node, or NULL on failure
+void *list_append(struct list *list, void *data)
+{
+    struct list_item *item = malloc(sizeof(struct list_item));
+    if (item == NULL) {
+        return NULL;
+    }
+    item->next = NULL;
+    item->data = data;
+    if (list->head == NULL) {
+        list->head = item;
+    } else {
+        list->tail->next = item;
```

Comment:
oh no we still have `tail`, but is it still necessary given the list is single now?